std-bits/table/src/main/java/org/enso/table/data/index/CrossTabIndex.java

```diff
@@ -0,0 +1,214 @@
+package org.enso.table.data.index;
+
+import java.util.*;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.enso.base.text.TextFoldingStrategy;
+import org.enso.table.aggregations.Aggregator;
+import org.enso.table.data.column.builder.Builder;
+import org.enso.table.data.table.Column;
+import org.enso.table.data.table.Table;
+import org.enso.table.error.TooManyColumnsException;
+import org.enso.table.problems.AggregatedProblems;
+import org.enso.table.util.NameDeduplicator;
+import org.graalvm.polyglot.Context;
+
+public class CrossTabIndex {
+  private static final int MAXIMUM_CROSS_TAB_COLUMN_COUNT = 10000;
+
+  private Column[] xColumns;
+
+  private Column[] yColumns;
+
+  private MultiValueIndex<UnorderedMultiValueKey> combinedIndex;
+
+  private ObjectNumberer<UnorderedMultiValueKey> xKeyNumberer;
+
+  private ObjectNumberer<UnorderedMultiValueKey> yKeyNumberer;
+
+  private UnorderedMultiValueKey[][] grid;
+
+  public CrossTabIndex(Column[] xColumns, Column[] yColumns, int tableSize) {
+    this.xColumns = xColumns;
+    this.yColumns = yColumns;
+
+    // Create combined index
+    Column combinedColumns[] =
+        Stream.concat(Arrays.stream(xColumns), Arrays.stream(yColumns)).toArray(Column[]::new);
+    combinedIndex =
+        MultiValueIndex.makeUnorderedIndex(
+            combinedColumns, tableSize, TextFoldingStrategy.unicodeNormalizedFold);
```

Comment:
I'm wondering if we should have a separate `makeCombinedIndex` which would work with a `CrossTabCombinedKey` - such special key can contain both `UnorderedMultiValueKey` subkeys for `x` and `y` and then compute its own hash by combining the hashes of the 2 subkeys.

Let's say we have X x columns and Y y columns, and N rows.

Now we are computing the hashes (X+Y)*N + X*N + Y*N times (because computing the subkey essentially re-dos the computation) - resulting in 2*(X+Y)*N hash computations.

If we do a 'combined' approach, we can get to X*N+Y*N+N = (X+Y+1)*N computations which is much less - for each row we just compute the hash for the x subkey, then y subkey and yield the resulting key to be a composition of these two hashes.

It also feels like it could be easier to understand - currently we are slightly (ab)using the MultiValueKey - whereas what we seem to really need is a CombinedKey holding 2 separate MultiValueKeys

Reply:
I implemented this approach, but it was actually slower, which surprised me. I'll go with the current solution, since this task needs to get closed before too long.

Reply:
Thanks for the comparison - the result is surprising indeed. But yes, let's proceed with the current state then.